### PR TITLE
[PYTHON] Introduce Json Statistics Report aligned with C++ version

### PR DESCRIPTION
**Pull request by PiotrKrzem**
### Details:
 - Added average, detailed, sorted json - formatted statistics reports
 - Added reports are aligned with C++ counterparts
 - Added a flag to Python that allows to store the report in json format

### Tickets:
 - 80796

### CODE REVIEW

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 1]
```diff
@@ -2,40 +2,57 @@
...
 from datetime import timedelta
+from io import TextIOWrapper
 import os
-import sys
 from enum import Enum
+import json
 import csv
+from typing import Dict, List, Tuple
 from .logging import logger
+import numpy as np
```
Comment (ilya-lavrenov) [inline tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py:12]: do we need to add extra dependencies to requirements txt files?

Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py:12]: All of them are already present in the default Python (except for numpy, but it's already a requirement for benchmark app)

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 2]
```diff
@@ -98,7 +152,74 @@ def dump_performance_counters_sorted(self, prof_sorted_info):
...
+            list: List[Dict[str, str]] = []
+            for profiling_info in prof_info_list:
+                for info in profiling_info:
+                    node_info_dict = {}
+                    node_info_dict['name'] = info.node_name
+                    node_info_dict['node_type'] = info.node_type
+                    node_info_dict['status'] = str(info.status)
+                    node_info_dict['real_time'] = f"{info.real_time / timedelta(milliseconds=1):.3f}"
+                    node_info_dict['cpu_time'] = f"{info.cpu_time / timedelta(milliseconds=1):.3f}"
+                    node_info_dict['exec_type'] = info.exec_type
```
Comment (jiwaszki) [inline tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py:198]:
```suggestion
                    node_info_dict = {
                        'name': info.node_name,
                        'node_type': info.node_type,
                        # ... rest of variables
                    }
```
It looks cleaner being initialized this way.

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 3]
```diff
@@ -107,30 +228,68 @@ def dump_performance_counters(self, prof_info_list):
...
+            json.dump(json_statistics, file, indent=4)
+
+    def dump_performance_counters_sorted(self, prof_sorted_info):
+        def profiling_info_to_dict_list(prof_info_matrix: np.ndarray):
+            total = 0
+            total_cpu = 0
+
+            list: List[Dict[str, str]] = []
+            for info in prof_info_matrix:
+                node_info_dict = {}
```
Comment (jiwaszki) [inline tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py:266]: Same here, initialization

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 4]
```diff
@@ -107,30 +228,68 @@ def dump_performance_counters(self, prof_info_list):
...
+            json_statistics = {}
+
+            nodes_info_list, total, total_cpu = profiling_info_to_dict_list(prof_sorted_info)
+
+            json_statistics['report_type'] = 'sorted'
+            json_statistics['avg_performance'] = {
+                'nodes': nodes_info_list,
+                'total_real_time': f"{total / 1000:.3f}",
+                'total_cpu_time': f"{total_cpu / 1000:.3f}"
+            }
```
Comment (jiwaszki) [inline tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py:293]:
```suggestion
            nodes_info_list, total, total_cpu = profiling_info_to_dict_list(prof_sorted_info)

            json_statistics = {
                'report_type': 'sorted',
                'avg_performance': {
                    # ...
                 }
            }
```

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 5]
```diff
@@ -98,7 +152,74 @@ def dump_performance_counters_sorted(self, prof_sorted_info):
...
+class JsonStatisticsReport(StatisticsReport):
+    def __init__(self, config):
+        StatisticsReport.__init__(self, config)
+
+    def dump(self):
+        def list_to_dict(parameters: List[Tuple[str, str]]) -> Dict[str, str]:
+            parameters_as_dict: Dict[str, str] = {}
+            for key, value in parameters:
+                parameters_as_dict[key] = value
+            return parameters_as_dict
```
Comment (jiwaszki) [inline tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py:164]:
```suggestion
        def list_to_dict(parameters: List[Tuple[str, str]]) -> Dict[str, str]:
            return {key: value for key, value in parameters}
```

**Review by jiwaszki — COMMENTED** [review]
General ideas to make it more pythonic

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: tools/benchmark_tool/openvino/tools/benchmark/main.py [hunk 6]
```diff
@@ -60,8 +61,11 @@ def main():
 
         command_line_arguments = get_command_line_arguments(sys.argv)
         if args.report_type:
-          statistics = StatisticsReport(StatisticsReport.Config(args.report_type, args.report_folder))
-          statistics.add_parameters(StatisticsReport.Category.COMMAND_LINE_PARAMETERS, command_line_arguments)
+            if args.json_stats:
+                statistics = JsonStatisticsReport(StatisticsReport.Config(args.report_type, args.report_folder))
+            else:
+                statistics = CsvStatisticsReport(StatisticsReport.Config(args.report_type, args.report_folder))
```
[the remark this answers is not in the document]
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/main.py]: Applied!

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 7]
```diff
@@ -59,21 +76,58 @@ def dump_parameters(f, parameters):
...
-            total += pi.real_time
-            total_cpu += pi.cpu_time
-        f.write(self.csv_separator.join(['Total', '', '', '',
-            f"{total / timedelta(milliseconds=1):.3f}",
-            f"{total_cpu / timedelta(milliseconds=1):.3f}"]))
-        f.write('\n\n')
+    def dump_performance_counters(self, prof_info_list):
+        def dump_performance_counters_request(self, f, prof_info):
+            total = timedelta()
+            total_cpu = timedelta()
```
[the remark this answers is not in the document]
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py]: Applied!

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py (hunk 2 above)
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py]: Agree, applied!

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py (hunk 5 above)
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py]: Applied!

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 8]
```diff
@@ -98,7 +152,74 @@ def dump_performance_counters_sorted(self, prof_sorted_info):
...
+        filename = os.path.join(self.config.report_folder, 'benchmark_report.json')
+        with open(filename, 'w') as file:
+            json_statistics = {}
+            if self.Category.COMMAND_LINE_PARAMETERS in self.parameters.keys():
+                json_statistics["cmd_options"] = \
+                    list_to_dict(self.parameters[self.Category.COMMAND_LINE_PARAMETERS])
+
+            if self.Category.RUNTIME_CONFIG in self.parameters.keys():
+                json_statistics["configuration_setup"] = \
+                    list_to_dict(self.parameters[self.Category.COMMAND_LINE_PARAMETERS])
```
[the remark this answers is not in the document]
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py]: Thank you for noticing! Applied!

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 9]
```diff
@@ -2,40 +2,57 @@
...
-        self.csv_separator = ';'
 
     def add_parameters(self, category, parameters):
         if category not in self.parameters.keys():
             self.parameters[category] = parameters
         else:
             self.parameters[category].extend(parameters)
 
+    def dump(self):
+        raise NotImplementedError("This function must be overwritten in the child class")
```
[the remark this answers is not in the document]
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py]: Honestly, I did not know before that it existed. Applied!

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py [hunk 10]
```diff
@@ -59,21 +82,59 @@ def dump_parameters(f, parameters):
...
-                f"{pi.cpu_time / timedelta(milliseconds=1):.3f}"]))
-            f.write('\n')
-            total += pi.real_time
-            total_cpu += pi.cpu_time
-        f.write(self.csv_separator.join(['Total', '', '', '',
-            f"{total / timedelta(milliseconds=1):.3f}",
-            f"{total_cpu / timedelta(milliseconds=1):.3f}"]))
-        f.write('\n\n')
+    def dump_performance_counters(self, prof_info_list):
+        def dump_performance_counters_request(self, f, prof_info):
```
Comment (Wovchena) [inline tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py:86]:
```suggestion
        def dump_performance_counters_request(f, prof_info):
```

File: tools/benchmark_tool/openvino/tools/benchmark/parameters.py [hunk 11]
```diff
@@ -163,6 +163,8 @@ def parse_args():
                            "counters and latency for each executed infer request.")
     stat.add_argument('-report_folder', '--report_folder', type=str, required=False, default='',
                       help="Optional. Path to a folder where statistics report is stored.")
+    args.add_argument('-json_stats', '--json_stats', type=str2bool, required=False, default=False, nargs='?', const=True,
+                      help="Optional. Enables JSON-based statistics output (by default reporting system will use CSV format). Should be used together with -report_folder option.")
```
Comment (Wovchena) [inline tools/benchmark_tool/openvino/tools/benchmark/parameters.py:167]: Running `../bin/intel64/Release/benchmark_app -niter 2 -m ~/.cache/omz/intel/face-detection-0200/FP16/face-detection-0200.xml -report_folder . -report_type detailed_counters -json_stats` saves `benchmark_detailed_counters_report.json` in which `detailed_performance` is a list of two big elements. Running `py ../tools/benchmark_tool/benchmark_app.py -niter 2 -m ~/.cache/omz/intel/face-detection-0200/FP16/face-detection-0200.xml -report_folder . -report_type detailed_counters -json_stats` saves only one big element into `detailed_performance` entry in `benchmark_detailed_counters_report.json`
Please align the behavior.

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py (hunk 10 above)
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/utils/statistics_report.py]: Applied!

File: tools/benchmark_tool/openvino/tools/benchmark/parameters.py (hunk 11 above)
Comment (PiotrKrzem) [reply tools/benchmark_tool/openvino/tools/benchmark/parameters.py:167]: Aligned!

**Review by nkogteva — APPROVED** [review]
LGTM

**Review by Wovchena — APPROVED** [review]

**Pull request merged**